### PR TITLE
Add DSR model

### PR DESCRIPTION
**Pull request by phcarval**
# Description

- Implementation of the DSR model, from https://github.com/VitjanZ/DSR_anomaly_detection 
- References #1121
- Architecture works without crashes (AFAIK), but might not yield the expected performance
- Pretraining needs weights : automatic weight download needs to be set up (unsure how, it's not on TIMM)
- Third phase of training not implemented yet (unsure of how to preceed)
- Still various pre-commit warnings flaring up

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

### CODE REVIEW

**Comment by phcarval** [comment]
Added automatic pretrained weight downloading. Now the model can train on its own, but the third phase of training is still missing.
These additions would benefit from being integrated to download.py (as well as the anomaly generator should be added to the already existing files... but this is more complex since DSR has its own way of generating anomalies).

**Comment by phcarval** [comment]
I tested the DSR model by loading the authors' pretrained weights and finding the exact same results as them. The following changes are necessary to reproduce :
- `load_state_dict()` for each module in `DsrModel.__init__()`
- normalization: none in config.yaml
- normalization_method: none in config.yaml
- eval batch size is 1
- comment `image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)` in data/utils/image.py
- set `weight_file` to None in `test.py`

Still training left to test as well as implementing the third phase of training.

**Comment by phcarval** [comment]
Training seems to work on my end. Only thing left to implement is the upsampling module.

**Comment by phcarval** [comment]
DSR is, AFAIK, fully functional right now. Training, testing is implemented for both phases of training. Training procedure from the paper has been copied (just need to adapt the nb of epochs or nb of steps in the config file). There are several small corrections that should be made now to clean up the code.

Let me know if there are problems on your end.

**Comment by phcarval** [comment]
I made it so that the scheduler can accept max_steps to anneal the learning rate. BTW, I believe you have an error in EfficientAD lightning_model.py L.188, since the StepLR scheduler's step_size argument sets the number of epochs: instead, you're giving it the number of actual steps. And I guess max() could pose a problem if you give positive max_steps and max_epochs, but one is supposed to be negative?

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for all the efforts! I have a few comments

File: src/anomalib/models/dsr/lightning_model.py
Comment (samet-akcay) [reply src/anomalib/models/dsr/lightning_model.py:68]: Yeah, when I run `python tools/train.py --model dsr`, I get the following error:

```bash
Traceback (most recent call last):
  File "<string>", line 2, in <module>
TypeError: len() takes exactly one argument (2 given)
```

File: src/anomalib/models/dsr/lightning_model.py
Comment (samet-akcay) [inline src/anomalib/models/dsr/lightning_model.py:70]:
```suggestion
        num_steps = max(
            self.trainer.max_steps // len(self.trainer.datamodule.train_dataloader()), self.trainer.max_epochs
        )
```

File: src/anomalib/models/dsr/loss.py
Comment (samet-akcay) [inline src/anomalib/models/dsr/loss.py:70]:
```suggestion
    def forward(self, pred_mask: Tensor, true_mask: Tensor) -> Tensor:
```

File: src/anomalib/models/dsr/lightning_model.py
Comment (phcarval) [reply src/anomalib/models/dsr/lightning_model.py:72]: I remember trying to move this to `on_train_start`. I think `configure_optimizers` is called first so this creates an issue?

**Review by samet-akcay — COMMENTED** [review]
@phcarval, thanks again for you effort. I've got few comments after my initial round of review, and a question:

I'm training the model now. However, the performance does not seem to be improving. For example, I'm on epoch 272 now, and the performance is the following since the first few epochs:

```
pixel_F1Score=0.109, pixel_AUROC=0.442
```

I see that the total number of epochs is set to 700. Is it improving by then? :)

File: src/anomalib/models/dsr/torch_model.py
Comment (phcarval) [reply src/anomalib/models/dsr/torch_model.py]: Done. Can't test it right now though, but I double checked everything.

File: src/anomalib/models/dsr/download_weights.py
Comment (phcarval) [reply src/anomalib/models/dsr/download_weights.py:2]: I'm going to refactor this, unless you add the weights to Anomalib?

File: src/anomalib/models/dsr/lightning_model.py
Comment (phcarval) [reply src/anomalib/models/dsr/lightning_model.py:68]: Yes indeed, did not test that commit, thanks.

File: src/anomalib/models/dsr/config.yaml
Comment (phcarval) [reply src/anomalib/models/dsr/config.yaml:31]: Sure. I did not know it worked that way, actually. I'm waiting to see if @ashwinvaidya17 adds the weights to Anomalib before changing it.

File: src/anomalib/models/dsr/download_weights.py
Comment (phcarval) [reply src/anomalib/models/dsr/download_weights.py:1]: Same as above.

**Comment by phcarval** [comment]
> @phcarval, thanks again for you effort. I've got few comments after my initial round of review, and a question:
> 
> I'm training the model now. However, the performance does not seem to be improving. For example, I'm on epoch 272 now, and the performance is the following since the first few epochs:
> 
> ```
> pixel_F1Score=0.109, pixel_AUROC=0.442
> ```
> 
> I see that the total number of epochs is set to 700. Is it improving by then? :)

Yes, it's normal since we're not training the upsampling module yet. Your values are actually very similar to mine.
image

However, you should see the image metrics improving.

**Comment by phcarval** [comment]
> @phcarval the weights are now part of a release https://github.com/openvinotoolkit/anomalib/releases/tag/dsr_pretrained_weights. Let me know if this works for you.

Great, thanks! However, we only need `checkpoints/vq_model_pretrained_128_4096.pckl` from the zip file. Could you change the release to only include that file? This would also allow me to use the already existing function `extract_and_download()`.

**Comment by ashwinvaidya17** [comment]
> > @phcarval the weights are now part of a release https://github.com/openvinotoolkit/anomalib/releases/tag/dsr_pretrained_weights. Let me know if this works for you.
> 
> Great, thanks! However, we only need `checkpoints/vq_model_pretrained_128_4096.pckl` from the zip file. Could you change the release to only include that file? This would also allow me to use the already existing function `extract_and_download()`.

I have updated the release. Lmk

**Comment by phcarval** [comment]
I am now using the existing download function. I have deleted `DsrWeightDownloader`. I have also made minor fixes to EfficientAD because now there are two methods using the `pre_trained/` folder. I have also fixed the epoch length problem for the EfficientAD optimizer, as it is similar to DSR.

**Comment by phcarval** [comment]
It turns out that if I set `max_steps` to 120000 for example in the config file, training stops much earlier than expected. On "bottle", training stops at 2222 epochs when it should run to 4444 epochs with a train batch size of 8. Thus, the upsampling module and the scheduler are not attained during training. The training time is basically divided by two. Could this be because of the fact that we use two optimizers, so it counts as two steps even though only one is effectively used at a time?

Note that this problem does not arise when using `max_epochs`.

**Comment by phcarval** [comment]
Another issue: the score output by DSR is not computed from the upsampled map, but from the original map. That means that if the upsampling module is performing poorly but the anomaly detection module works well, we will have a very good image-level score but a very poor pixel-level score. When looking at the results folder, it becomes impossible to see which images were actually correctly classified and which images just show poor segmentation performance.

**Comment by phcarval** [comment]
Added and tested batch evaluation (`max` -> `amax`). For now, everything OK but the step count and the actual decision not showing in the output images.

**Comment by samet-akcay** [comment]
@ashwinvaidya17, @djdameln can we prioritise this please? It's been outstanding for a bit here. Thanks for your patience @phcarval

**Comment by phcarval** [comment]
Hi, did you have some time to review the PR? :)
I have merged the main branch. The issues I mentioned still exist but I'm waiting on your input to fix them.

**Review by ashwinvaidya17 — APPROVED** [review]
Sorry for the late review. Thanks again for the efforts!

**Comment by samet-akcay** [comment]
Thanks @phcarval, I'll have a look at the issues. Just to double check, is your description up-to-date? I want to know if you have addressed any of the issues you mentioned there.

**Comment by phcarval** [comment]
> Thanks @phcarval, I'll have a look at the issues. Just to double check, is your description up-to-date? I want to know if you have addressed any of the issues you mentioned there.

Oh no actually, I didn't update it. The two issues I still are the following :


> It turns out that if I set `max_steps` to 120000 for example in the config file, training stops much earlier than expected. On "bottle", training stops at 2222 epochs when it should run to 4444 epochs with a train batch size of 8. Thus, the upsampling module and the scheduler are not attained during training. The training time is basically divided by two. Could this be because of the fact that we use two optimizers, so it counts as two steps even though only one is effectively used at a time?
> 
> Note that this problem does not arise when using `max_epochs`.



> Another issue: the score output by DSR is not computed from the upsampled map, but from the original map. That means that if the upsampling module is performing poorly but the anomaly detection module works well, we will have a very good image-level score but a very poor pixel-level score. When looking at the results folder, it becomes impossible to see which images were actually correctly classified and which images just show poor segmentation performance.
> 
> I would suggest adding the classification results to each output image. Of course, for most methods it would show redundant information that is technically already present through the output segmentation mask.

**Comment by phcarval** [comment]
Readded the path check that was removed in commit d2b6e5a

**Comment by phcarval** [comment]
@samet-akcay have you had the time to look at the issues?

**Comment by samet-akcay** [comment]
> @samet-akcay have you had the time to look at the issues?

I was out of office, will check this out this week. I'll let you know if I have any update

**Comment by blaz-r** [comment]
Hello @phcarval, this all seems like really nice work 😄. 
About your issues: one thing I noticed in the [lightning documentation](https://pytorch-lightning.readthedocs.io/en/1.4.9/common/optimizers.html#automatic-optimization) is that optimizer.step is indeed called for every optimizer (looks like even if None is returned but I'm not sure):
```python
for epoch in epochs:
    for batch in data:
        for opt in optimizers:

            def closure():
                loss = model.training_step(batch, batch_idx, optimizer_idx)
                opt.zero_grad()
                loss.backward()
                return loss

            opt.step(closure)

    for lr_scheduler in lr_schedulers:
        lr_scheduler.step()
```
And based on this https://github.com/Lightning-AI/lightning/issues/16143 I'd say that count goes up for every step call.
That's probably why you see half of max_steps.
[Later in the docs](https://pytorch-lightning.readthedocs.io/en/1.4.9/common/optimizers.html#step-optimizers-at-arbitrary-intervals) they mention that if you want to optimize at custom times, you should override `optimizer_step()` function. I think this might be a proper solution in this case.

Regarding the other thing about the classification and segmentation discrepancy. From the paper it seems that for some cases (transistor) the difference might be quite significant, and I'm just interested if you have any results (metrics and visuals) on that?
From what you say, and also for other cases, I do think that the visualization should also include the anomaly score (which I think actually is somewhat included in some inferencers like OpenVINO). This would probably need a separate PR.

**Comment by phcarval** [comment]
Hi @blaz-r ! Thanks a lot for the comments.

The `optimizer_step()` function indeed looks interesting. I guess it does advance one step for each optimizer, even if returning None, so I'll look into using that.

I have no specific results to show: rather, this problem stems from the optimizer issue. If, for some reason, someone does not train the upsampling module for enough steps, they should be able to see which images were correctly classified. In that case, I agree, another PR should be issued for that.

**Comment by phcarval** [comment]
I have the same problem as https://github.com/Lightning-AI/lightning/discussions/10893: I cannot make an "empty" optimizer step, so the step count increases for each optimizer.

**Comment by phcarval** [comment]
I fixed it using manual optimization. However, it doesn't improve maintainability and perhaps it should be changed if we find a way to make it work using core Lightning functionality.

**Comment by blaz-r** [comment]
Nice 😄. That's actually a good solution, in my opinion even a more clean one. Good job.
We discussed the visualization, and the idea is to add a label with the image name and anomaly score to the current visualizer. That would of course be a separate PR. Does that sound like something you had in mind?

**Comment by phcarval** [comment]
> Nice 😄. That's actually a good solution, in my opinion even a more clean one. Good job. 

Thanks for the feedback!

> We discussed the visualization, and the idea is to add a label with the image name and anomaly score to the current visualizer. That would of course be a separate PR. Does that sound like something you had in mind?

Indeed, this proposition sounds fine to me.

**Comment by samet-akcay** [comment]
> I fixed it using manual optimization. However, it doesn't improve maintainability and perhaps it should be changed if we find a way to make it work using core Lightning functionality.

Thanks for the fix @phcarval! I agree with @blaz-r. Your solution also shows that this could be done manually via vanilla torch, which demonstrates flexibility. I think it would be a good example to some users/developers for future implementations.

**Comment by phcarval** [comment]
> > I fixed it using manual optimization. However, it doesn't improve maintainability and perhaps it should be changed if we find a way to make it work using core Lightning functionality.
> 
> Thanks for the fix @phcarval! I agree with @blaz-r. Your solution also shows that this could be done manually via vanilla torch, which demonstrates flexibility. I think it would be a good example to some users/developers for future implementations.

Thanks for the feedback! My concern was that, AFAIK, it is the only method that uses a different optimization system in Anomalib and so it could create somewhat of an issue down the line if the method is refactored, as optimization is supposed to be "automatic" in Lightning. It's probably a very minor problem though.

Anyway, I have no more outstanding issues on my end!

File: src/anomalib/models/dsr/torch_model.py
Comment (samet-akcay) [inline src/anomalib/models/dsr/torch_model.py:28]: Is this intentionally left blank?

Comment (samet-akcay) [reply src/anomalib/models/dsr/torch_model.py:28]: Also, as raised above, `anom_par` is not so clear to understand. It would be great if it is renamed

File: src/anomalib/models/dsr/torch_model.py
Comment (samet-akcay) [inline src/anomalib/models/dsr/torch_model.py:107]: left over?

File: src/anomalib/models/dsr/torch_model.py
Comment (samet-akcay) [inline src/anomalib/models/dsr/torch_model.py:109]: It would be more readable if `outputs` is annotated. Something like 
```suggestion
        outputs: dict[str, Tensor]
```

File: src/anomalib/models/dsr/torch_model.py
Comment (samet-akcay) [inline src/anomalib/models/dsr/torch_model.py:206]: is `anomaly_map` really a `true_mask`?

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Thanks @phcarval, this is a huge PR :)

I've got some minor comments, but the overall code quality is quite high! Thanks again for your contribution and patience!

**Comment by samet-akcay** [comment]
By the way, there is one failed test, mainly because `DsrModel` returns a `dict`. We probably need a separate PR to handle torch outputs that are dict

File: docs/source/reference_guide/algorithms/dsr.rst
Comment (phcarval) [reply docs/source/reference_guide/algorithms/dsr.rst:16]: The link doesn't work as of now, is this intended?

File: src/anomalib/models/dsr/README.md
Comment (phcarval) [reply src/anomalib/models/dsr/README.md:13]: As above

File: src/anomalib/models/dsr/config.yaml
Comment (phcarval) [reply src/anomalib/models/dsr/config.yaml:31]: I suggest `anomaly_strength` or `latent_anomaly_strength`

File: src/anomalib/models/dsr/lightning_model.py
Comment (phcarval) [reply src/anomalib/models/dsr/lightning_model.py:53]: Will do

File: src/anomalib/models/dsr/torch_model.py
Comment (phcarval) [reply src/anomalib/models/dsr/torch_model.py:28]: See above

File: docs/source/reference_guide/algorithms/dsr.rst
Comment (samet-akcay) [reply docs/source/reference_guide/algorithms/dsr.rst:16]: it will work once we merge this

**Comment by phcarval** [comment]
I went though all requested changes @samet-akcay !

File: src/anomalib/models/dsr/lightning_model.py
Comment (samet-akcay) [inline src/anomalib/models/dsr/lightning_model.py:170]: Is it possible if `model_outputs` have the same keys for `anomaly_maps` and `pred_scores`? The reason for asking is that we need standard keys for the `TorchInferencer` stuff?

The tests are failing since the output model is returning a different output than what is expected here
https://github.com/openvinotoolkit/anomalib/blob/5eb20c64cc6ee8862bf2723b83fc4cae60a5594e/src/anomalib/deploy/inferencers/torch_inferencer.py#L155-L185

Comment (samet-akcay) [reply src/anomalib/models/dsr/lightning_model.py:170]: I have create PR #1354 to handle cases like this. `DsrModel` needs to return `anomaly_map` and `pred_score` though.

**Comment by phcarval** [comment]
@samet-akcay have you had the time to test the model? I seem to have very bad segmentation results, this might be due to the optimizer change. I would like to know if you had that problem too.

**Comment by phcarval** [comment]
When removing the default imagenet normalization parameter, training seems to be going much better and my results are close to the paper's. Only problem is the one I describe in issue https://github.com/openvinotoolkit/anomalib/issues/1380 where images are segmented even if they're classified as non-anomalous.

**Review by samet-akcay — APPROVED** [review]
Thanks @phcarval, looking great! Really appreciate your patience, and thanks again for this great contribution!

**Pull request merged**
